src/Api.Rest/Common/Utilities/OAuthTokenValidator.cs

```diff
@@ -0,0 +1,91 @@
+﻿#region copyright
+
+/* * * * * * * * * * * * * * * * * * * * * * * * * */
+/* Carl Zeiss Industrielle Messtechnik GmbH        */
+/* Softwaresystem PiWeb                            */
+/* (c) Carl Zeiss 2024                             */
+/* * * * * * * * * * * * * * * * * * * * * * * * * */
+
+#endregion
+
+namespace Zeiss.PiWeb.Api.Rest.Common.Utilities;
+
+#region usings
+
+using System;
+using System.Collections.Generic;
+using System.IdentityModel.Tokens.Jwt;
+using System.Linq;
+using System.Security.Claims;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using IdentityModel;
+using IdentityModel.Client;
+using Microsoft.IdentityModel.Tokens;
+using Zeiss.PiWeb.Api.Rest.HttpClient.OAuth;
+
+#endregion
+
+/// <summary>
+/// Provides different methods to validate identity token.
+/// </summary>
+public static class OAuthTokenValidator
+{
+	#region methods
+
+	/// <summary>
+	/// Checks if the provided nonce in the token claims is valid and matching the expected nonce.
+	/// </summary>
+	/// <param name="expectedNonce">The expected nonce.</param>
+	/// <param name="tokenClaims">A collection of claims.</param>
+	/// <returns><see langword="true"/> if nonce is valid, otherwise <see langword="false"/>.</returns>
+	public static bool ValidateNonce( string expectedNonce, IEnumerable<Claim> tokenClaims )
+	{
+		var tokenNonce = tokenClaims.FirstOrDefault( c => c.Type == JwtClaimTypes.Nonce );
+
+		return tokenNonce != null && string.Equals( tokenNonce.Value, expectedNonce, StringComparison.Ordinal );
+	}
+
+	/// <summary>
+	/// Checks if the code hash in the claims is valid.
+	/// </summary>
+	/// <param name="authorizationCode">The acquired authorization code from the identity provider.</param>
+	/// <param name="tokenClaims">The claims of the acquired identity token.</param>
+	/// <returns><see langword="true"/> if code hash is valid, otherwise <see langword="false"/>.</returns>
+	public static bool ValidateCodeHash( string authorizationCode, IEnumerable<Claim> tokenClaims )
+	{
+		var cHash = tokenClaims.FirstOrDefault( c => c.Type == JwtClaimTypes.AuthorizationCodeHash );
+
+		using var sha = SHA256.Create();
+
+		var codeHash = sha.ComputeHash( Encoding.ASCII.GetBytes( authorizationCode ) );
+		var leftBytes = new byte[ 16 ];
+		Array.Copy( codeHash, leftBytes, 16 );
```

Comment:
Currently we can expect it because we create the token ourself, but it makes sense to be future proof. I changed this method to use the correct algorithm based on the token. We support all algorithms specified here: https://datatracker.ietf.org/doc/html/rfc7518#section-3.1